ql/src/java/org/apache/hadoop/hive/ql/DriverTxnHandler.java

```diff
@@ -527,11 +527,15 @@ void release() {
     release(!hiveLocks.isEmpty());
   }
 
-  void destroy() {
+  void destroy(String queryIdFromDriver) {
+    // We need cleanup transactions, even if we did not acquired locks yet
+    // However TxnManager is bound to session, so wee need to check if it is already handling a new query
```

Comment:
If the TxnManager is handling a new query, would the old transaction be closed correctly?

Reply:
If txnManager is handling a new query, the previous one must be either committed or aborted otherwise the second query's openTxn call would have failed with an exception.